server/src/test/java/org/elasticsearch/index/shard/GlobalCheckpointListenersTests.java

```diff
@@ -0,0 +1,402 @@
+        final int numberOfListeners = randomIntBetween(0, 16);
+        final boolean[] failures = new boolean[numberOfListeners];
+        final IndexShardClosedException[] exceptions = new IndexShardClosedException[numberOfListeners];
+        for (int i = 0; i < numberOfListeners; i++) {
+            final int index = i;
+            final boolean failure = randomBoolean();
+            failures[index] = failure;
+            final GlobalCheckpointListeners.GlobalCheckpointListener listener =
+                    (g, e) -> {
+                        assert g == UNASSIGNED_SEQ_NO;
+                        assert e != null;
+                        if (failure) {
+                            throw new RuntimeException("failure");
+                        } else {
+                            exceptions[index] = e;
+                        }
+                    };
+            globalCheckpointListeners.add(NO_OPS_PERFORMED, listener);
+        }
+        globalCheckpointListeners.close();
+        for (int i = 0; i < numberOfListeners; i++) {
+            if (failures[i]) {
+                assertNull(exceptions[i]);
+            } else {
+                assertNotNull(exceptions[i]);
+                assertThat(exceptions[i].getShardId(), equalTo(shardId));
+            }
+        }
+        int failureCount = 0;
+        for (int i = 0; i < numberOfListeners; i++) {
+            if (failures[i]) {
+                failureCount++;
+            }
+        }
+        if (failureCount > 0) {
+            final ArgumentCaptor<String> message = ArgumentCaptor.forClass(String.class);
+            final ArgumentCaptor<RuntimeException> t = ArgumentCaptor.forClass(RuntimeException.class);
+            verify(mockLogger, times(failureCount)).warn(message.capture(), t.capture());
+            assertThat(message.getValue(), equalTo("error notifying global checkpoint listener of closed shard"));
+            assertNotNull(t.getValue());
+            assertThat(t.getValue().getMessage(), equalTo("failure"));
+        }
+    }
+
+    public void testNotificationUsesExecutor() {
+        final AtomicInteger count = new AtomicInteger();
+        final Executor executor = command -> {
+            count.incrementAndGet();
+            command.run();
+        };
+        final GlobalCheckpointListeners globalCheckpointListeners = new GlobalCheckpointListeners(shardId, executor, logger);
+        globalCheckpointListeners.globalCheckpointUpdated(NO_OPS_PERFORMED);
+        final int numberOfListeners = randomIntBetween(0, 16);
+        for (int i = 0; i < numberOfListeners; i++) {
+            globalCheckpointListeners.add(NO_OPS_PERFORMED, (g, e) -> {});
+        }
+        globalCheckpointListeners.globalCheckpointUpdated(randomLongBetween(NO_OPS_PERFORMED, Long.MAX_VALUE));
+        assertThat(count.get(), equalTo(1));
+    }
+
+    public void testConcurrency() throws BrokenBarrierException, InterruptedException {
```

Comment:
Can you also add closing to this test? So that each listener gets notified exactly once, even under concurrent closing.

Reply:
Good idea, this caught an issue. I pushed 9533679a44b40a87446f4bacb2188e0ce426e10d.